File: decompiled/c/5e65cf.c

/* WARNING: Globals starting with '_' overlap smaller symbols at the same address */

undefined4 FUN_005e65cf(void)

{
  byte bVar1;
  undefined4 in_EAX;
  undefined4 uVar2;
  undefined4 uVar3;
  undefined4 extraout_ECX;
  char unaff_BL;
  uint uVar4;
  int iVar5;
  undefined4 unaff_ESI;
  uint uVar6;
  uint uVar7;
  bool bVar8;
  undefined8 uVar9;
  byte *pbVar10;
  
  if ((_DAT_0099a500 & 1) != 0) {
    return in_EAX;
  }
  uVar9 = FUN_00431510();
  pbVar10 = (byte *)((ulonglong)uVar9 >> 0x20);
  uVar3 = (undefined4)uVar9;
  if (unaff_BL == '\x03') {
    if ((*pbVar10 & 0x3c) == 4) {
      return uVar3;
    }
    uVar4 = (uint)pbVar10[7];
    iVar5 = uVar4 * 0x260;
    if ((&DAT_00887441)[iVar5] == '\0') {
      DAT_005f54f0 = 0x4c1;
      if ((*pbVar10 & 0x3c) == 0x10) {
        DAT_005f54f2 = 0x56b;
        if (pbVar10[4] != 0) {
          DAT_005f54f2 = 0x56d;
        }
      }
      else {
        bVar1 = pbVar10[4];
        if (((bVar1 != 2) && (bVar1 != 3)) && (bVar1 != 1)) {
          DAT_005f54f2 = (&DAT_00887442)[uVar4 * 0x130];
          _DAT_005f54f4 = (undefined2)(&DAT_00887444)[uVar4 * 0x98];
          _DAT_005f54f6 = (undefined2)((uint)(&DAT_00887444)[uVar4 * 0x98] >> 0x10);
          return uVar3;
        }
        DAT_005f54f2 = 0x569;
      }
      if (1 < (byte)(&DAT_00887497)[iVar5]) {
        DAT_005f54f2 = DAT_005f54f2 + 1;
      }
      _DAT_005f54f4 = (&DAT_00887442)[uVar4 * 0x130];
      _DAT_005f54f6 = (undefined2)(&DAT_00887444)[uVar4 * 0x98];
      _DAT_005f54f8 = (undefined2)((uint)(&DAT_00887444)[uVar4 * 0x98] >> 0x10);
      _DAT_005f54fa = *(short *)(&DAT_005f5806 + (uint)(byte)(&DAT_00887420)[iVar5] * 8) + 2;
      uVar6 = (*(ushort *)(pbVar10 + 5) & 0x70) >> 4;
      uVar7 = uVar6;
      do {
        if ((&DAT_0088744a)[uVar4 * 0x130 + uVar6] == -1) {
          uVar7 = uVar7 - 1;
        }
        uVar6 = uVar6 - 1;
      } while (-1 < (int)uVar6);
      _DAT_005f54fc = (short)uVar7 + 1;
      return uVar3;
    }
  }
  else if (unaff_BL == '\x02') {
    if (*pbVar10 != 0) {
      return uVar3;
    }
    uVar4 = (uint)pbVar10[0x30];
    if ((&DAT_00887441)[uVar4 * 0x260] == '\0') {
      DAT_005f54f0 = 0x4c1;
      DAT_005f54f2 = (&DAT_00887442)[uVar4 * 0x130];
      _DAT_005f54f4 = (undefined2)(&DAT_00887444)[uVar4 * 0x98];
      _DAT_005f54f6 = (undefined2)((uint)(&DAT_00887444)[uVar4 * 0x98] >> 0x10);
      return uVar3;
    }
  }
  else {
    bVar8 = (DAT_00991f30 & 0x48) == 0x48;
    if (!bVar8) {
      uVar3 = extraout_ECX;
      uVar2 = FUN_005e3b2b();
      uVar9 = CONCAT44(pbVar10,uVar2);
      if (bVar8) {
        uVar3 = FUN_005e3b2b(unaff_ESI,pbVar10,uVar3);
        uVar9 = CONCAT44(pbVar10,uVar3);
        if (bVar8) {
          return uVar3;
        }
      }
    }
    uVar4 = (uint)((ulonglong)uVar9 >> 0x20);
    uVar3 = (undefined4)uVar9;
    if (unaff_BL == '\x05') {
      DAT_005f54f0 = 0x4c2;
      DAT_005f54f2 = *(undefined2 *)(&DAT_006e1ecc + (uint)*(byte *)(uVar4 + 4) * 8);
      return uVar3;
    }
    if (unaff_BL == '\x06') {
      DAT_005f54f0 = 0x4c2;
      DAT_005f54f2 = (*(byte *)(uVar4 + 4) >> 4) + 0x4d1;
      return uVar3;
    }
    if (unaff_BL == '\a') {
      DAT_005f54f0 = 0x4c2;
      DAT_005f54f2 = (*(byte *)(uVar4 + 5) & 0xf) + 0x632;
      return uVar3;
    }
    if (unaff_BL == '\t') {
      DAT_005f54f0 = 0x4c2;
      DAT_005f54f2 = *(undefined2 *)(&DAT_0099fc3c + (uVar4 >> 8 & 0xff) * 2);
      return uVar3;
    }
    if (unaff_BL == '\n') {
      DAT_005f54f0 = 0x4c2;
      DAT_005f54f2 = (*(ushort *)(uVar4 + 4) & 0x3ff) + 0x7de;
      return uVar3;
    }
  }
  return uVar3;
}

File: ported/auto/5e65cf.js
// Auto-translated from Ghidra C by tools/c-to-js/translate.js.
// Source: decompiled/c/5e65cf.c
// Edit by hand only after diff-test passes — re-running the translator will overwrite.

/** @typedef {import("../../runtime/heap.js").Heap} Heap */

import { CONCAT44 } from "../../runtime/ghidra-builtins.js";
import { regs } from "../../runtime/regs.js";
import { FUN_00431510 } from "./431510.js";
import { FUN_005e3b2b } from "./5e3b2b.js";
export function FUN_005e65cf(heap) {
  let bVar1 = 0;
  let in_EAX = regs.eax >>> 0;
  let uVar2 = 0;
  let uVar3 = 0;
  let extraout_ECX = 0;
  let unaff_BL = regs.ebx & 0xff;
  let uVar4 = 0;
  let iVar5 = 0;
  let unaff_ESI = regs.esi >>> 0;
  let uVar6 = 0;
  let uVar7 = 0;
  let bVar8 = 0;
  let uVar9 = 0;
  let pbVar10 = 0;
  if ((heap.u32(0x0099a500) & 1) != 0) {
    return in_EAX;
  }
  uVar9 = (((regs.edx = 0x9, regs.eax = FUN_00431510(heap))) >>> 0);
  pbVar10 = (((((uVar9) >>> 0) >>> 0x20)) >>> 0);
  uVar3 = ((((uVar9) >>> 0)) >>> 0);
  if (unaff_BL == 3) {
    if ((heap.u8(pbVar10) & 0x3c) == 4) {
      return uVar3;
    }
    uVar4 = ((((heap.u8(pbVar10 + (7))) >>> 0)) >>> 0);
    iVar5 = ((uVar4 * 0x260) >>> 0);
    if (heap.u32((0x00887441) + (iVar5) * 4) == 0) {
      heap.setU32(0x005f54f0, (0x4c1) >>> 0);
      if ((heap.u8(pbVar10) & 0x3c) == 0x10) {
        heap.setU32(0x005f54f2, (0x56b) >>> 0);
        if (heap.u8(pbVar10 + (4)) != 0) {
          heap.setU32(0x005f54f2, (0x56d) >>> 0);
        }
      } else {
        bVar1 = ((heap.u8(pbVar10 + (4))) & 0xff);
        if (((bVar1 != 2) && (bVar1 != 3)) && (bVar1 != 1)) {
          heap.setU32(0x005f54f2, (heap.u32((0x00887442) + (uVar4 * 0x130) * 4)) >>> 0);
          heap.setU32(0x005f54f4, (heap.u32(((0x00887444) & 0xffff) + (uVar4 * 0x98) * 4)) >>> 0);
          heap.setU32(0x005f54f6, (((heap.u32(((0x00887444) >>> 0) + (uVar4 * 0x98) * 4) >>> 0x10) & 0xffff)) >>> 0);
          return uVar3;
        }
        heap.setU32(0x005f54f2, (0x569) >>> 0);
      }
      if (1 < heap.u32(((0x00887497) & 0xff) + (iVar5) * 4)) {
        heap.setU32(0x005f54f2, (heap.u32(0x005f54f2) + 1) >>> 0);
      }
      heap.setU32(0x005f54f4, (heap.u32((0x00887442) + (uVar4 * 0x130) * 4)) >>> 0);
      heap.setU32(0x005f54f6, (heap.u32(((0x00887444) & 0xffff) + (uVar4 * 0x98) * 4)) >>> 0);
      heap.setU32(0x005f54f8, (((heap.u32(((0x00887444) >>> 0) + (uVar4 * 0x98) * 4) >>> 0x10) & 0xffff)) >>> 0);
      heap.setU32(0x005f54fa, (heap.i16((0x005f5806 + heap.u32(((0x00887420) >>> 0) + (iVar5) * 4) * 8)) + 2) >>> 0);
      uVar6 = (((heap.u16((pbVar10 + 5)) & 0x70) >>> 4) >>> 0);
      uVar7 = ((uVar6) >>> 0);
      do {
        if ((heap.u32((0x0088744a) + (uVar4 * 0x130 + uVar6) * 4) | 0) == -1) {
          uVar7 = ((uVar7 - 1) >>> 0);
        }
        uVar6 = ((uVar6 - 1) >>> 0);
      } while (-1 < (((uVar6) | 0) | 0));
      heap.setU32(0x005f54fc, (((uVar7) << 16 >> 16) + 1) >>> 0);
      return uVar3;
    }
  } else {
    if (unaff_BL == 2) {
    if (heap.u8(pbVar10) != 0) {
      return uVar3;
    }
    uVar4 = ((((heap.u8(pbVar10 + (0x30))) >>> 0)) >>> 0);
    if (heap.u32((0x00887441) + (uVar4 * 0x260) * 4) == 0) {
      heap.setU32(0x005f54f0, (0x4c1) >>> 0);
      heap.setU32(0x005f54f2, (heap.u32((0x00887442) + (uVar4 * 0x130) * 4)) >>> 0);
      heap.setU32(0x005f54f4, (heap.u32(((0x00887444) & 0xffff) + (uVar4 * 0x98) * 4)) >>> 0);
      heap.setU32(0x005f54f6, (((heap.u32(((0x00887444) >>> 0) + (uVar4 * 0x98) * 4) >>> 0x10) & 0xffff)) >>> 0);
      return uVar3;
    }
  } else {
    bVar8 = (((heap.u32(0x00991f30) & 0x48) == 0x48) & 0xff);
    if (!bVar8) {
      uVar3 = ((extraout_ECX) >>> 0);
      uVar2 = (((regs.ecx = 0x8d, regs.edx = 0x9, regs.eax = FUN_005e3b2b(heap))) >>> 0);
      uVar9 = ((CONCAT44(pbVar10, uVar2)) >>> 0);
      if (bVar8) {
        uVar3 = (((regs.ecx = 0x8d, regs.edx = 0x9, regs.eax = FUN_005e3b2b(heap, unaff_ESI, pbVar10, uVar3))) >>> 0);
        uVar9 = ((CONCAT44(pbVar10, uVar3)) >>> 0);
        if (bVar8) {
          return uVar3;
        }
      }
    }
    uVar4 = ((((((uVar9) >>> 0) >>> 0x20) >>> 0)) >>> 0);
    uVar3 = ((((uVar9) >>> 0)) >>> 0);
    if (unaff_BL == 5) {
      heap.setU32(0x005f54f0, (0x4c2) >>> 0);
      heap.setU32(0x005f54f2, (heap.u16((0x006e1ecc + heap.u32((uVar4 + 4)) * 8))) >>> 0);
      return uVar3;
    }
    if (unaff_BL == 6) {
      heap.setU32(0x005f54f0, (0x4c2) >>> 0);
      heap.setU32(0x005f54f2, ((heap.u8((uVar4 + 4)) >>> 4) + 0x4d1) >>> 0);
      return uVar3;
    }
    if (unaff_BL == 7) {
      heap.setU32(0x005f54f0, (0x4c2) >>> 0);
      heap.setU32(0x005f54f2, ((heap.u8((uVar4 + 5)) & 0xf) + 0x632) >>> 0);
      return uVar3;
    }
    if (unaff_BL == 9) {
      heap.setU32(0x005f54f0, (0x4c2) >>> 0);
      heap.setU32(0x005f54f2, (heap.u16((0x0099fc3c + (uVar4 >>> 8 & 0xff) * 2))) >>> 0);
      return uVar3;
    }
    if (unaff_BL == 10) {
      heap.setU32(0x005f54f0, (0x4c2) >>> 0);
      heap.setU32(0x005f54f2, ((heap.u16((uVar4 + 4)) & 0x3ff) + 0x7de) >>> 0);
      return uVar3;
    }
  }
  }
  return uVar3;
}
